### PR TITLE
Mod rewrite and new models

### PR DESCRIPTION
**Pull request by nxet**
This PR introduces a completely redesigned API for the mod, allowing for a streamlined generation of the chest definition with stronger generic abstractions of the underlying methods.
All models have been rewritten as well, adapting to the new layout, and two new models have been added (Fridge and Toolbox).
Updated the localization template and added Italian localization.

I'm not 100% sure pipeworks compatibility has been correctly ported, some testing would be much appreciated.

### CODE REVIEW

File: utils/base.lua [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+			local meta = minetest.get_meta(pos)
+			local inv = meta:get_inventory()
+			return inv:is_empty("main")
+		end,
+	}
+	-- register log actions, NOTE passing an anonymous function to avoid getting the default if not necessary
+	out.allow_metadata_inventory_move = parse_action(def.allow_metadata_inventory_move, function() actions.get_allow_metadata_inventory_move{def.type} end)
+	out.allow_metadata_inventory_put = parse_action(def.allow_metadata_inventory_put, function() actions.get_allow_metadata_inventory_put{def.type} end)
+	out.allow_metadata_inventory_take = parse_action(def.allow_metadata_inventory_take, function() actions.get_allow_metadata_inventory_take{def.type} end)
+	out.on_metadata_inventory_move = parse_action(def.on_metadata_inventory_move, function() actions.get_on_metadata_inventory_move(def.type) end)
```
Comment (SmallJoker) [inline utils/base.lua:63]: Which of those? `{def.type}` or `(def.type)`?

File: utils/formspec.lua [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    cfg.chest_width .. "," .. cfg.chest_height .. ";]" ..
+  	"list[current_player;main;" ..
+    player_inv_x_orig .. "," .. player_inv_y_orig ..
+    ";8,1;]" ..
+  	"list[current_player;main;" ..
+    player_inv_x_orig .. "," .. (player_inv_y_orig + 1.15) ..
+    ";8,3;8]" ..
+  	"listring[current_name;main]" ..
+  	"listring[current_player;main]" ..
+  	default.get_hotbar_bg(player_inv_x_orig, player_inv_y_orig)
```
Comment (SmallJoker) [inline utils/formspec.lua:35]: Mixed tabs/spaces.

**Comment by SmallJoker** [comment]
This is indeed a drastic change. In all cases, backwards compatibility must be kept to not break maps:

 * same node names
 * pipeworks works (testing needed)
 * if there was external API: try to re-implement

This is just a general information. Yet I did neither test nor look at the PR in detail.

File: utils/base.lua (hunk 1 above)
Comment (nxet) [reply utils/base.lua:63]: Actually both are valid. As you can see in actions.lua the `get_allow_*` methods are designed in order to set a default argument in the local metatable, while the `get_on_*` don't need such functionality

File: utils/formspec.lua (hunk 2 above)
Comment (nxet) [reply utils/formspec.lua:35]: Apologies for that, will be fixed in a coming commit.

**Comment by nxet** [comment]
I definitely did work with backwards compatibility in mind, and all* the elements of the mod had their public attributes kept identical to those on the master branch.
The only difference is in the recipes, which now have a more generic ingredient requirement (default:wood -> group:wood).
* As stated in the PR I unfortunately have no means to properly test the pipeworks compatibility, and would appreciate the effort and help of members of the community in order to check out all possible issues with my implementation.

I'd like to point out I literally started 2 days ago getting into Minetest, modding and the Lua language, so there might be many oversights on my part. I cannot stress enough how useful it would be to involve members of the community to properly test this update.

**Comment by SmallJoker** [comment]
FYI: I left a message for @Megaf on IRC to check whether he is okay with these major changes. So far no answer.

**Comment by nxet** [comment]
Thanks a lot, I have been wondering what was happening with this PR and I’m
glad to know it hasn’t been forgotten. Looking forward any news, thanks
again!

**Comment by louisroyer** [comment]
Hi @nxet, could you please also update the `README.md` with your new models description?

**Comment by louisroyer** [comment]
Also, I put here the french translation for your new nodes. Feel free to add them.
```text
Big Fridge=Grand réfrigérateur
Fridge=Réfrigérateur

Acacia Wood Toolbox=Boîte à outils en acacia
Aspen Wood Toolbox=Boîte à outils en tremble
Junglewood Toolbox=Boîte à outils en bois de la jungle
Pine Wood Toolbox=Boîte à outils en pin
Steel Toolbox=Boîte à outils en acier
Wooden Toolbox=Boîte à outils en bois
```

File: models/shared.lua [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	tiles = {
+		top = "shared_top.png",
+		side = "shared_side.png",
+		front = "shared_front.png"
+	},
+	formspec = get_formspec(""),
+	pipeworks_enabled = true,
+	sounds = default.node_sound_wood_defaults(),
+	recipe = {
+		{"group:wood", "default:leaves", "group:wood"},
```
Comment (louisroyer) [inline models/shared.lua:48]:
```suggestion
		{"group:wood", "group:leaves", "group:wood"},
```

File: README.md [hunk 4]
```diff
@@ -1,47 +1,85 @@
...
-===========
+<style>
+    th { display: none; }
+    td { text-align: center; }
+</style>
 
-More Chests
+# more_chests
+Originally a fork of 0gb.us's chests_0gb_us (https://forum.minetest.net/viewtopic.php?f=11&t=4366).  
+Megaf's more_chests fixes several bugs, uses new textures and adds compatibility with [VanessaE's Pipeworks] (https://github.com/VanessaE/pipeworks) mod.
```
Comment (louisroyer) [inline README.md:8]:
```suggestion
Originally a fork of [0gb.us's chests_0gb_us](https://forum.minetest.net/viewtopic.php?f=11&t=4366).  
Megaf's more_chests fixes several bugs, uses new textures and adds compatibility with [VanessaE's Pipeworks](https://github.com/VanessaE/pipeworks) mod.  
```

File: README.md [hunk 5]
```diff
@@ -1,47 +1,85 @@
-more_chests
-===========
+<style>
+    th { display: none; }
+    td { text-align: center; }
+</style>
```
Comment (louisroyer) [inline README.md:4]:
```suggestion

```
Github flavored markdown doesn't support this.

File: README.md [hunk 6]
```diff
@@ -1,47 +1,85 @@
...
-{'default:wood','default:cobble','default:wood'},
-{'default:cobble','default:steel_ingot','default:cobble'},
-{'default:wood','default:cobble','default:wood'}
+#### Cobble Chest
+This locked chest looks like cobblestone, and has no info text. Great for hiding things in. However, unlike real cobblestone, this chest is breakable by hand. If you suspect there is one hiding, hold the left mouse button and run your hand along the walls. When cracks appear, you've found the chest.
+1 | 2 | 3
+---|---|---
+default:wood | default:cobble | default:wood
+default:cobble | default:steel_ingot | default:cobble
+default:wood | default:cobble | default:wood
```
Comment (louisroyer) [inline README.md:21]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>group:wood</td><td>default:cobble</td><td>group:wood</td>
        <tr><td>default:cobble</td><td>default:steel_ingot</td><td>default:cobble</td>
        <tr><td>group:wood</td><td>default:cobble</td><td>group:wood</td>
    </tbody>
</table>
```

File: README.md [hunk 7]
```diff
@@ -1,47 +1,85 @@
...
-{'default:wood','default:steel_ingot','default:wood'},
-{'default:wood','default:wood','default:wood'}
 
+#### Drop Box
 Anyone can put things in, but only the chest's placer can remove items.
+1 | 2 | 3
+---|---|---
+default:wood | _empty_ | default:wood
+default:wood | default:steel_ingot | default:wood
+default:wood | default:wood | default:wood
```
Comment (louisroyer) [inline README.md:30]:
```suggestion

<table>
    <tbody align="center">
        <tr><td>group:wood</td><td><em>empty</em></td><td>group:wood</td>
        <tr><td>group:wood</td><td>default:steel_ingot</td><td>group:wood</td>
        <tr><td>group:wood</td><td>group:wood</td><td>group:wood</td>
    </tbody>
</table>
```

File: README.md [hunk 8]
```diff
@@ -1,47 +1,85 @@
...
-{'default:wood','default:wood','default:wood'}
 
-As long as you remember to click “close” before you leave the chest, no one can see what the chest contains. Only the chest's owner can click “open” and “close” on the chest's formspec, revealing and hiding the chest's contents.
+#### Secret Chest
+As long as you remember to click "close" before you leave the chest, no one can see what the chest contains. Only the chest's owner can click "open" and "close" on the chest's formspec, revealing and hiding the chest's contents.
+1 | 2 | 3
+---|---|---
+default:wood | default:cobble | default:wood
+default:wood | default:steel_ingot | default:wood
+default:wood | default:wood | default:wood
```
Comment (louisroyer) [inline README.md:39]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>group:wood</td><td>default:cobble</td><td>group:wood</td>
        <tr><td>group:wood</td><td>default:steel_ingot</td><td>group:wood</td>
        <tr><td>group:wood</td><td>group:wood</td><td>group:wood</td>
    </tbody>
</table>
```

File: README.md [hunk 9]
```diff
@@ -1,47 +1,85 @@
...
-{'default:wood','default:steel_ingot','default:wood'},
-{'default:wood','default:wood','default:wood'}
 
+#### Shared Chest
 Exactly what it sounds like. The chest's placer can add people to the chest's shared list using the chest's formspec. Warning: anyone you add may empty the chest. When the chest is empty, it can be mined by anyone, just like a regular locked chest.
+1 | 2 | 3
+---|---|---
+default:wood | default:leaves | default:wood
+default:wood | default:steel_ingot | default:wood
+default:wood | default:wood | default:wood
```
Comment (louisroyer) [inline README.md:48]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>group:wood</td><td>group:leaves</td><td>group:wood</td>
        <tr><td>group:wood</td><td>default:steel_ingot</td><td>group:wood</td>
        <tr><td>group:wood</td><td>group:wood</td><td>group:wood</td>
    </tbody>
</table>
```

File: README.md [hunk 10]
```diff
@@ -1,47 +1,85 @@
...
-{'default:wood','default:wood','default:wood'}
 
+#### Wifi Chest
 A wacky chest that doesn't store it's items in the usual way, but instead, stores them remotely. For that reason, all wifi chests appear to have the same inventory. Due to not actually having an inventory, wifi chests can also be mined, even when they appear to have stuff in them. Lastly, as everyone gets their own wifi account, the items you see in the wifi chest are not the same items anyone else sees. This chest's properties make it nice for keeping secrets, as well as essentially almost doubling your inventory space, if you choose to carry one with you.
-```
+1 | 2 | 3
+---|---|---
+default:wood | default:mese | default:wood
+default:wood | default:steel_ingot | default:wood
+default:wood | default:wood | default:wood
```
Comment (louisroyer) [inline README.md:57]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>group:wood</td><td>default:mese</td><td>group:wood</td>
        <tr><td>group:wood</td><td>default:steel_ingot</td><td>group:wood</td>
        <tr><td>group:wood</td><td>group:wood</td><td>group:wood</td>
    </tbody>
</table>
```

File: README.md [hunk 11]
```diff
@@ -1,47 +1,85 @@
...
+
+#### Fridge
+A new model which comes in two forms, 1- and 2-block tall, which indeed have different sizes inventories. No fancy functionality here, I just wanted to know at a glance which one is the chest I'm storing food into.
+
+###### Fridge Recipe
+1 | 2 | 3
+---|---|---
+_empty_ | default:steel_ingot | _empty_
+default:steel_ingot | default:ice | default:steel_ingot
+_empty_ | default:steel_ingot | _empty_
```
Comment (louisroyer) [inline README.md:68]:
```suggestion
<table>
    <tbody align="center">
        <tr><td><em>empty</em></td><td>default:steel_ingot</td><td><em>empty</em></td>
        <tr><td>default:steel_ingot</td><td>default:ice</td><td>default:steel_ingot</td>
        <tr><td><em>empty</em></td><td>default:steel_ingot</td><td><em>empty</em></td>
    </tbody>
</table>
```

File: README.md [hunk 12]
```diff
@@ -1,47 +1,85 @@
...
+
+###### Big Fridge Recipe
+1 | 2 | 3
+---|---|---
+default:steel_ingot | default:steel_ingot | default:steel_ingot
+default:steel_ingot | default:ice | default:steel_ingot
+default:steel_ingot | default:steel_ingot | default:steel_ingot
```
Comment (louisroyer) [inline README.md:75]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>default:steel_ingot</td><td>default:steel_ingot</td><td>default:steel_ingot</td>
        <tr><td>default:steel_ingot</td><td>default:ice</td><td>default:steel_ingot</td>
        <tr><td>default:steel_ingot</td><td>default:steel_ingot</td><td>default:steel_ingot</td>
    </tbody>
</table>
```

File: README.md [hunk 13]
```diff
@@ -1,47 +1,85 @@
...
+
+
+#### Toolbox
+This model has no particular functionality to offer other than giving your tools' chest a new look. As a bonus you can also craft the chest with different types of wood, which will give you different results. For the fanciest garage workshop you ever built on Minetest.
+1 | 2 | 3
+---|---|---
+default:wood | default:wood | default:wood
+default:wood | default:pickaxe | default:wood
+default:wood | default:wood | default:wood
```
Comment (louisroyer) [inline README.md:84]:
```suggestion
<table>
    <tbody align="center">
        <tr><td>material</td><td>material</td><td>material<td>
        <tr><td>material</td><td>group:pickaxe</td><td>material</td>
        <tr><td>material</td><td>material</td><td>material</td>
    </tbody>
</table>
```

File: README.md [hunk 14]
```diff
@@ -1,47 +1,85 @@
...
+> Note: crafting also accepts Aspen, Acacia, Junglewood, Pine and Steel instead of wood
```
Comment (louisroyer) [inline README.md:85]:
```suggestion
> Note: replace `material` with either `default:wood`, `default:aspen_wood`, `default:acacia_wood`, `default:junglewood`, `default:pine_wood`, or `default:steel_ingot`.
```

**Review by louisroyer — CHANGES_REQUESTED** [review]
Github flavored markdown doesn't support `<style>`, here is a fix.
Also, I suggest using `group:leaves` instead of `leaves`.

File: README.md (hunk 6 above)
Comment (SmallJoker) [reply README.md:21]: This table is valid markdown, though some parsers might prefer padded contents like this:

```
1            | 2              | 3
-------------|----------------|-------------
default:wood | default:cobble | default:wood
```
But I really don't know. The current version is rendered correctly:
https://github.com/minetest-mods/more_chests/blob/52cabc9fbc695a7eb86238e21ec6b6e0ae431c5b/README.md

Comment (louisroyer) [reply README.md:21]: Yes, it is valid, but there is `<style> th { display: none; } td { text-align: center; } </style>` at the top of the rendered page, and it is not supposed to be rendered this way (it is supposed to hide the header, and to center the content of the table instead, and this is not the case).

What I see with the current version:
image

image

**Comment by Megaf** [comment]
Thank you so much for the time and work you put into this!
Thanks @nxet @SmallJoker @louisroyer

File: README.md (hunk 6 above)
Comment (S-S-X) [reply README.md:21]: Just my personal opinion as I stumbled upon this PR: README.md files should not contain HTML or CSS and files should be readable with simple editors that do not render HTML tables for example like `vim` or similar cli editors.
Formatting suggested by @SmallJoker imo is also very good for all use cases.

**Pull request merged**